File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
'''
- 정렬 문제
- 정렬기준 : max 재생 장르, max 재생, min 고유번호
- 정렬 값 : index

1. 장르별 재생횟수 합산
2. 배열 : [장르재생횟수, 재생횟수, 고유번호]
'''

def solution(genres, plays):
    answer = []
    gen_map = {}
    
    # 장르별 재생횟수 합산
    for i, g in enumerate(genres):
        gen_map[g] = gen_map.setdefault(g, 0) + plays[i]
    
    # [장르재생횟수, 재생횟수, 고유번호] 배열 생성
    arr = []
    for i in range(len(plays)):
        arr.append([gen_map[genres[i]], plays[i], i])
    
    # 기준별 정렬
    arr = sorted(arr, key=lambda x: (-x[0], -x[1], x[2]))
    
    # 각 장르별로 최대 2개씩 담기
    best_arr = [arr[0][2]]
    before, before_idx = arr[0], 1
    for i in range(1, len(arr)):
        if before[0] != arr[i][0]:
            best_arr.append(arr[i][2])
            before = arr[i]
            before_idx = 1
        elif before_idx == 1:
            best_arr.append(arr[i][2])
            before_idx += 1
            
    return best_arr

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        int[] answer = {};
        HashMap<String, Integer> gen_map = new HashMap<>();
        for(int i = 0; i < plays.length; i++) {
            gen_map.put(genres[i], gen_map.getOrDefault(genres[i], 0) + plays[i]);
        }
        int[][] arr = new int[plays.length][3];
        
        for(int i = 0; i < plays.length; i++) {
            arr[i] = new int[] {gen_map.get(genres[i]), plays[i], i};
        }
        
        Arrays.sort(arr, (n1, n2) -> {
            if(n1[0] == n2[0]) {
                if(n1[1] == n2[1]) {
                    return Integer.compare(n1[2], n2[2]);
                }
                return Integer.compare(n2[1], n1[1]);
            } else {
                return Integer.compare(n2[0], n1[0]);
            }
        });
        
        ArrayList<Integer> best_arr = new ArrayList<>();
        
        best_arr.add(arr[0][2]);
        int[] before = arr[0];
        int before_cnt = 1;
        for(int i = 1; i < arr.length; i++) {
            if(before[0] != arr[i][0]) {
                best_arr.add(arr[i][2]);
                before = arr[i];
                before_cnt = 1;
            } else if(before_cnt == 1) {
                best_arr.add(arr[i][2]);
                before_cnt += 1;
            }
        }
        answer = new int[best_arr.size()];
        for(int i = 0; i < answer.length; i++) {
            answer[i] = best_arr.get(i);
        }
        return answer;
    }
}